File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] edge) {
        ArrayList<Integer>[] list;
        list = new ArrayList[n+1];
        
        for(int i=0; i<=n; i++){
            list[i] = new ArrayList<>();
        }
        
        for(int i=0; i<edge.length; i++){
            int s = edge[i][0];
            int e = edge[i][1];
            
            list[s].add(e);
            list[e].add(s);
        }
        
        int[] dist = new int[n+1];
        boolean[] visited = new boolean[n+1];
        
        ArrayDeque<Integer> q = new ArrayDeque<>();
        q.add(1);
        visited[1] = true;
        
        while(!q.isEmpty()){
            int now = q.poll();
            
            for(int next : list[now]){
                if(!visited[next]){
                    visited[next] = true;
                    dist[next] = dist[now] + 1;
                    q.add(next);
                }
            }
        }
        
        int max = Integer.MIN_VALUE;
        for(int i=1; i<=n; i++){
            max = Math.max(max, dist[i]);
        }
        
        int answer = 0;
        for(int i=1; i<=n; i++){
            if(dist[i] == max) answer++;
        }
        return answer;
    }
}

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.py
from collections import deque
graph = [[]]
def solution(n, edge):
    global graph
    graph = [[] for _ in range(n+1)]
    dist = [-1] * (n+1)
    for s, e in edge:
        graph[s].append(e)
        graph[e].append(s)
    
    bfs(1, dist)
    maxDist = max(dist)
    
    answer = 0
    for i in range(n+1):
        if dist[i] == maxDist:
            answer += 1
    return answer

def bfs(start, dist):
    global graph
    q = deque()
    dist[start] = 0
    q.append(start)
    
    while q:
        now = q.popleft()
        
        for next in graph[now]:
            if dist[next] == -1:
                dist[next] = dist[now] + 1
                q.append(next)
    